File: 264-ugly-number-ii/ugly-number-ii.py
import heapq
class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        pq=[]
        heapq.heappush(pq,1)
        seen=set()
        seen.add(1)
        while 1:
            cur=heapq.heappop(pq)
            n-=1
            if n==0:
                return cur
            
            if cur*2 not in seen:
                seen.add(cur*2)
                heapq.heappush(pq,cur*2)
            if cur*3 not in seen:
                seen.add(cur*3)
                heapq.heappush(pq,cur*3)
            if cur*5 not in seen:
                seen.add(cur*5)
                heapq.heappush(pq,cur*5)
        

File: 264-ugly-number-ii/ugly-number-ii.cpp
class Solution {
public:
    int nthUglyNumber(int n) {
        priority_queue <long long,vector <long long>, greater <long long>> pq;
        pq.push(1);
        map <long long,bool> mp;
        while(1){
            long long cur=pq.top();
            pq.pop();
            n--;
            if(n==0){
                return cur;
            }
            if(mp.find(cur*2)==mp.end()){
                pq.push(cur*2);
                mp[cur*2]=true;
            }
            if(mp.find(cur*3)==mp.end()){
                pq.push(cur*3);
                mp[cur*3]=true;
            }
            if(mp.find(cur*5)==mp.end()){
                pq.push(cur*5);
                mp[cur*5]=true;
            }
        }
    }
};